snapshots/handler/handler.go

```diff
@@ -0,0 +1,156 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package handler
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/opencontainers/go-digest"
+	"github.com/opencontainers/image-spec/identity"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+const (
+	TargetSnapshotLabel = "containerd.io/snapshot.ref"
+	TargetRefLabel      = "containerd.io/snapshot/target.reference"
+	TargetDigestLabel   = "containerd.io/snapshot/target.digest"
+)
+
+// FilterLayerBySnapshotter filters out layers from download candidates if the
+// snapshotter can prepare the layer snapshot without pulling it.
+func FilterLayerBySnapshotter(f images.HandlerFunc, sn snapshots.Snapshotter, store content.Store, fetcher remotes.Fetcher, ref string) images.HandlerFunc {
+	return func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		children, err := f(ctx, desc)
+		if err != nil {
+			return nil, err
+		}
+
+		if desc.MediaType == ocispec.MediaTypeImageManifest ||
+			desc.MediaType == images.MediaTypeDockerSchema2Manifest {
+			p, err := content.ReadBlob(ctx, store, desc)
+			if err != nil {
+				return nil, err
+			}
+			var manifest ocispec.Manifest
+			if err := json.Unmarshal(p, &manifest); err != nil {
+				return nil, err
+			}
+
+			if _, err := remotes.FetchHandler(store, fetcher)(ctx, manifest.Config); err != nil {
+				return nil, err
+			}
+			chain, err := images.RootFS(ctx, store, manifest.Config)
+			if err != nil {
+				return nil, err
+			}
+
+			// Get the index of a layer which is the lowest non-skippable layer. We avoid
+			// downloading skippable layers which are lower than it by filtering out them.
+			i := firstFalseIdx(manifest.Layers, chain, isPullSkippable(ctx, sn, ref))
+			children = exclude(children, manifest.Layers[:i])
+		}
+
+		return children, nil
+	}
+}
+
+func isPullSkippable(ctx context.Context, sn snapshots.Snapshotter, ref string) func(ocispec.Descriptor, []digest.Digest) bool {
```

Comment:
Out of scope thinking: containerd provides runtime function and there were two copies of image data: content and snapshotter. For running containers without building function, content data seems duplicate. We can remove the content data after unpack and the `isPullSkippable` can avoid next round pulling. For example, image A is based on image B. We download image B first and remove the layers from image B. And then when we pulling image A, we can only pull extra layers from A.

Based on your commits, no matter what kind of snapshot is, remote or local snapshoter, if backend snapshotter follow the #3793 design, we can remove duplicate content data.